engine/api/websocket.go

```diff
@@ -0,0 +1,330 @@
+				return
+			}
+		case <-tick.C:
+			msg, err := store.GetMessageFromSubscription(ctx, pubSub)
+			if err != nil {
+				log.Warning(ctx, "websocketBroker.cacheSubscribe> Cannot get message %s: %s", msg, err)
+				continue
+			}
+			var e sdk.Event
+			if err := json.Unmarshal([]byte(msg), &e); err != nil {
+				// don't print the error as we doesn't care
+				continue
+			}
+
+			observability.Record(b.router.Background, WebSocketEvents, 1)
+			cacheMsgChan <- e
+		}
+	}
+}
+
+func (b *websocketBroker) ServeHTTP() service.Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (err error) {
+
+		c, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			log.Warning(ctx, "websocket> upgrade: %v", err)
+			return err
+		}
+		defer c.Close()
+
+		client := websocketClient{
+			UUID:             sdk.UUID(),
+			AuthConsumer:     getAPIConsumer(ctx),
+			isAlive:          abool.NewBool(true),
+			con:              c,
+			updateFilterChan: make(chan sdk.WebsocketFilter, 10),
+		}
+		b.chanAddClient <- &client
+
+		sdk.GoRoutine(ctx, fmt.Sprintf("readUpdateFilterChan-%s-%s", client.AuthConsumer.GetUsername(), client.UUID), func(ctx context.Context) {
+			client.readUpdateFilterChan(ctx, b.dbFunc())
+		})
+
+		for {
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+
+			var msg sdk.WebsocketFilter
+			_, message, err := c.ReadMessage()
+			if err != nil {
+				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+					log.Warning(ctx, "websocket error: %v", err)
+				}
+				log.Debug("%s disconnected", client.AuthConsumer.GetUsername())
+				break
+			}
+
+			if err := json.Unmarshal(message, &msg); err != nil {
+				log.Warning(ctx, "websocket.readJSON: %v", err)
+			}
+
+			// Send message to client
+			client.updateFilterChan <- msg
+		}
+		return nil
+	}
+}
+
+func (c *websocketClient) readUpdateFilterChan(ctx context.Context, db *gorp.DbMap) {
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("events.Http: context done")
+			return
+		case m := <-c.updateFilterChan:
+			if err := c.updateEventFilter(ctx, db, m); err != nil {
+				log.Error(ctx, "websocketClient.readUpdateFilterChan: unable to update event filter: %v", err)
+				msg := sdk.WebsocketEvent{
+					Status: "KO",
+					Error:  sdk.Cause(err).Error(),
+				}
+				_ = c.con.WriteJSON(msg)
+				continue
+			}
+		}
+	}
+}
+
+func (c *websocketClient) updateEventFilter(ctx context.Context, db gorp.SqlExecutor, m sdk.WebsocketFilter) error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// Subscribe to project
+	if (m.ProjectKey != "" && m.WorkflowName == "") || (m.ProjectKey != "" && m.Operation != "") {
+		perms, err := permission.LoadProjectMaxLevelPermission(ctx, db, []string{m.ProjectKey}, getAPIConsumer(ctx).GetGroupIDs())
+		if err != nil {
+			return err
+		}
+		maxLevelPermission := perms.Level(m.ProjectKey)
+		if maxLevelPermission < sdk.PermissionRead && !isMaintainer(ctx) && !isAdmin(ctx) {
```

Comment:
For now there is nothing that prevent to create a consumer for a hatchery that is maintainer or admin. But is this case we don't want to send all event but only ones based on groups ids.